nvme_tcp.go

```diff
@@ -0,0 +1,632 @@
+			return Device{}, errors.New("connectSingleDevice canceled")
+		default:
+		}
+		devices = readDevicesFromResultCH(devCH, devices)
+		// check all discovery gorutines finished
+		if !discoveryComplete {
+			select {
+			case <-wgCH:
+				discoveryComplete = true
+				logger.Info(ctx, "all discovery goroutines complete")
+			default:
+				logger.Info(ctx, "discovery goroutines are still running")
+			}
+		}
+		if discoveryComplete && len(devices) == 0 {
+			msg := "discovery complete but devices not found"
+			logger.Error(ctx, msg)
+			return Device{}, errors.New(msg)
+		}
+		if wwn == "" && len(devices) != 0 {
+			var err error
+			wwn, err = c.scsi.GetDeviceWWN(ctx, devices)
+			if err != nil {
+				logger.Info(ctx, "wwn for devices %s not found", devices)
+			}
+		}
+		if wwn != "" {
+			for _, d := range devices {
+				if err := c.scsi.WaitUdevSymlink(ctx, d, wwn); err == nil {
+					logger.Info(ctx, "registered device found: %s", d)
+					return Device{Name: d, WWN: wwn}, nil
+				}
+			}
+		}
+		if discoveryComplete && !lastTry {
+			logger.Info(ctx, "discovery finished, wait %f seconds for device registration",
+				c.waitDeviceRegisterTimeout.Seconds())
+			lastTry = true
+			endTime = time.Now().Add(c.waitDeviceRegisterTimeout)
+		}
+		if lastTry && time.Now().After(endTime) {
+			msg := "registered device not found"
+			logger.Error(ctx, msg)
+			return Device{}, errors.New(msg)
+		}
+		time.Sleep(time.Second)
+	}
+}
+
+//inprogress
+func (c *NVMeTCPConnector) connectMultipathDevice(
+	ctx context.Context, sessions []gonvme.NVMESession, info NVMeTCPVolumeInfo) (Device, error) {
+	defer tracer.TraceFuncCall(ctx, "NVMeTCPConnector.connectMultipathDevice")()
+	devCH := make(chan string, len(sessions))
+	wg := sync.WaitGroup{}
+	discoveryCtx, cFunc := context.WithTimeout(ctx, c.waitDeviceTimeout)
+	defer cFunc()
+
+	wg.Add(1)
+	go c.discoverDevice(discoveryCtx, &wg, devCH, info)
+	// for non blocking wg wait
+	wgCH := make(chan struct{})
+	go func() {
+		wg.Wait()
+		close(wgCH)
+	}()
+
+	var devices []string
+	var wwn, mpath string
+	var wwnAdded, discoveryComplete, lastTry bool
+	var endTime time.Time
+	for {
+		// get discovered devices
+		select {
+		case <-ctx.Done():
+			return Device{}, errors.New("connectMultipathDevice canceled")
+		default:
+		}
+		devices = readDevicesFromResultCH(devCH, devices)
+		// check all discovery gorutines finished
+		if !discoveryComplete {
+			select {
+			case <-wgCH:
+				discoveryComplete = true
+				logger.Info(ctx, "all discover goroutines complete")
+			default:
+				logger.Info(ctx, "discover goroutines are still running")
+			}
+		}
+		if discoveryComplete && len(devices) == 0 {
+			msg := "discover complete but devices not found"
+			logger.Error(ctx, msg)
+			return Device{}, errors.New(msg)
+		}
+		if wwn == "" && len(devices) != 0 {
+			var err error
+			wwn, err = c.scsi.GetDeviceWWN(ctx, devices)
+			if err != nil {
+				logger.Info(ctx, "wwn for devices %s not found", devices)
```

Comment:
In some places log will be debug since it is actually not an error situation and we will not be returning as we will continue with discover iterations